src/interrupt/mod.rs

```diff
@@ -0,0 +1,193 @@
+//! activities.
+//!
+//! Hardware interrupts are used by devices to communicate that they require attention from the
+//! operating system, or a bare-metal program running on the CPU if there are no OSes. The act of
+//! initiating a hardware interrupt is referred to as an interrupt request (IRQ). Different devices
+//! are usually associated with different interrupts using a unique value associated with each
+//! interrupt. This makes it possible to know which hardware device caused which interrupts.
+//! These interrupt values are often called IRQ lines, or just interrupt lines.
+//!
+//! Nowadays, IRQ lines is not the only mechanism to delivery device interrupts to processors.
+//! MSI [(Message Signalled Interrupt)](https://en.wikipedia.org/wiki/Message_Signaled_Interrupts)
+//! is another commonly used alternative in-band method of signaling an interrupt, using special
+//! in-band messages to replace traditional out-of-band assertion of dedicated interrupt lines.
+//! While more complex to implement in a device, message signaled interrupts have some significant
+//! advantages over pin-based out-of-band interrupt signaling. Message signaled interrupts are
+//! supported in PCI bus since its version 2.2, and in later available PCI Express bus. Some non-PCI
+//! architectures also use message signaled interrupts.
+//!
+//! On the other hand, IRQ is a term commonly used by Operating Systems when dealing with hardware
+//! interrupts. The IRQ number managed by OSes is independent of the IRQ number managed by VMM.
+//! To make things easy, `Interrupt Source` is used instead of IRQ for pin-based interrupts and
+//! MSI interrupts.
+//!
+//! A device may support multiple types of interrupts, and each type of interrupt may support one
+//! or multiple interrupt sources. For example, a PCI device may support:
+//! * Legacy Irq: exactly one interrupt source.
+//! * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+//! * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+//!
+//! A distinct Interrupt Source Identifier (ISID) will be assigned to each interrupt source.
+//! An ID allocator will be used to allocate and free Interrupt Source Identifiers for devices.
+//! To decouple the vm-device crate from the ID allocator, the vm-device crate doesn't take the
+//! responsibility to allocate/free Interrupt Source IDs but only makes use of assigned IDs.
+//!
+//! The overall flow to deal with interrupts is:
+//! * the VMM creates an interrupt manager
+//! * the VMM creates a device manager, passing on an reference to the interrupt manager
+//! * the device manager passes on an reference to the interrupt manager to all registered devices
+//! * guest kernel loads drivers for virtual devices
+//! * guest device driver determines the type and number of interrupts needed, and update the
+//!   device configuration
+//! * the virtual device backend requests the interrupt manager to create an interrupt group
+//!   according to guest configuration information
```

Comment:
@jiangliu Thanks for the proposal.

The idea to make vm-device crate manage the interrupt resource and related methods is great.
One place not clear to me is, what does it mean by "device manager passes on an reference to the interrupt manager"? Do you mean that the device manager records the interrupt group that a device uses, e.g. `<LegacyIrq, Arc<Mutex<<dyn Device>>>`?

Reply:
That means passing an argument of Arc<Box<dyn InterruptManager>> to the device manager, so the device manager could clone/call methods of InterruptManager trait object.

Reply:
Thanks. 
Do we also have request to let device manager get the information of which interrupt/number/group a device uses?
Some future request like dump a device information or something similar...

Reply:
Aha, one point is missing in the proposal, how to allocate interrupt source identifier from the vm_allocator. 
I think pre-allocator enough interrupt source identifier for each device should work for most cases. For example, PCI MSIx supports 2048 interrupts, but 32 interrupts should work for most cases:)
Or we could determine number of interrupts needed when creating virtual device backend.

Reply:
For pin irq, virtual backend could specify a number or just let vm-allocator allocate one. For MSI/MSI-X, why we need preallocate?

Reply:
>For pin irq, virtual backend could specify a number or just let vm-allocator allocate one. 
Per my understanding, any interrupt source id allocated from the vm-allocator should be OK.

>For MSI/MSI-X, why we need preallocate?
Otherwise when guest OS enabling MSI, the virtual device backend needs to allocate interrupt source IDs from the allocator, which forces dependency on the vm-allocator.

Reply:
> > For MSI/MSI-X, why we need preallocate?
> > Otherwise when guest OS enabling MSI, the virtual device backend needs to allocate interrupt source IDs from the allocator, which forces dependency on the vm-allocator.

Not sure if there're some misunderstandings. :)
The guest is the only role to allocate MSI vector for a device. And as we talked offline, there's no need to set back to allocator when virtual device backend traps a MSI data (vector inside). So there'll be no dependency.

Reply:
By default, the kvm kernel module supports 1024 interrupt sources at most. For MSI, we still need to allocate a distinct interrupt source ID for for each MSI interrupt, so we could call KVM_SET_IRQ_ROUTING for setup interrupt route.

Reply:
Well we have two options to make the dependency between KVM and MSI. Either we store MSI vector so that they can be called directly with `KVM_SIGNAL_MSI`, or we decide some more generic design relying on `KVM_SET_GSI_ROUTING` and using `EventFd` to notify about an interrupt and let KVM do the mapping between the file descriptor and the associated MSI.
For instance, in the context of VFIO, `KVM_SET_GSI_ROUTING` is very convenient since it avoids the interrupt generation coming from the device to go back to userspace. But on the other hand, `KVM_SET_GSI_ROUTING` has a drawback, the entire table has to be updated everytime an MSI vector is updated.

Reply:
I think KVM_SET_GSI_ROUTING is the right way. KVM_SIGNAL_MSI is convenient for the VMM to inject interrupts into the guests, but it doesn't support posted interrupt and will cause penalty when dealing device pass through.
For KVM_SET_GSI_ROUTING, it depends on a global view of all interrupts, so we let the InterruptManager to deal with KVM_SET_GSI_ROUTING:)

Reply:
>For KVM_SET_GSI_ROUTING, it depends on a global view of all interrupts, so we let the InterruptManager to deal with KVM_SET_GSI_ROUTING:)

Easy to say :) and hopefully not too hard to put in practice. I'm only concerned about the fact that the `InterruptManager` will have to be aware about any MSI update coming from the guest. And those updates can come through BAR or PCI config writing, which means we need mechanisms to callback into the `InterruptManager` to let it know that something happened.
The part where the `InterruptManager` decides to update `KVM_SET_GSI_ROUTING` list based on its internal entries seems easy indeed :)

Reply:
@jiangliu @sboeuf 

> > For KVM_SET_GSI_ROUTING, it depends on a global view of all interrupts, so we let the InterruptManager to deal with KVM_SET_GSI_ROUTING:)
> 
> Easy to say :) and hopefully not too hard to put in practice. I'm only concerned about the fact that the `InterruptManager` will have to be aware about any MSI update coming from the guest. And those updates can come through BAR or PCI config writing, which means we need mechanisms to callback into the `InterruptManager` to let it know that something happened.
> The part where the `InterruptManager` decides to update `KVM_SET_GSI_ROUTING` list based on its internal entries seems easy indeed :)

So we decide to *only* make vm-device::InterruptManager to know the global view of all interrupts (especially msi udpate), instead of have vm-allocator to know that? 

So for interrupt, can I say that the actual work for vm-allocator now is to just pre-allocate enough for each device during initialization?

Reply:
That sounds right to me. Since the vm-allocator will never have to handle conflicts and/or allocations of MSI/MSI-X, the vm-device can be the one being notified about the values of those interrupt vectors only.

Reply:
> > For instance, in the context of VFIO, KVM_SET_GSI_ROUTING is very convenient since it avoids the interrupt generation coming from the device to go back to userspace. But on the other hand, KVM_SET_GSI_ROUTING has a drawback, the entire table has to be updated everytime an MSI vector is updated.
> 
> Agreed. That's a drawback, but I think `KVM_SET_GSI_ROUTING` is a more generic scheme that allows us to support both VMM injection and posted interrupts.

We are prototyping by using KVM_SET_GSI_ROUTING because it's very friendly to POSTED_INTR.

Reply:
It turns out current design works perfectly:)

Reply:
> It turns out current design works perfectly:)

You mean the KVM_SET_GSI_ROUTING based design? That's great, yes.